### PR TITLE
TRUS-2679: Amend Personal Rep

### PR DESCRIPTION
**Pull request by BriWak**

### CODE REVIEW

File: app/uk/gov/hmrc/estates/repositories/CacheRepository.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    } yield createdLastUpdatedIndex && createdIdIndex
+  }
+
+  override def get(utr: String, internalId: String): Future[Option[JsValue]] = {
+
+    val selector = Json.obj(
+      "id" -> createKey(utr, internalId)
+    )
+
+    collection.flatMap {collection =>
```
Comment (adamconder) [inline app/uk/gov/hmrc/estates/repositories/CacheRepository.scala:78]: Suspiscious shadowing here with the same variable names.

File: app/uk/gov/hmrc/estates/repositories/VariationsTransformationRepository.scala [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    } yield createdLastUpdatedIndex && createdIdIndex
+  }
+
+  override def get(utr: String, internalId: String): Future[Option[ComposedDeltaTransform]] = {
+
+    val selector = Json.obj(
+      "id" -> createKey(utr, internalId)
+    )
+
+    collection.flatMap {collection =>
```
Comment (adamconder) [inline app/uk/gov/hmrc/estates/repositories/VariationsTransformationRepository.scala:79]: Suspicious shadowing for value names

File: app/uk/gov/hmrc/estates/services/DesService.scala [hunk 3]
```diff
@@ -38,8 +43,33 @@ class DesService @Inject()(val desConnector: DesConnector) {
     desConnector.getSubscriptionId(trn)
   }
 
-  def getEstateInfo(utr: String): Future[GetEstateResponse] = {
-    desConnector.getEstateInfo(utr)
+  def refreshCacheAndGetTrustInfo(utr: String, internalId: String)(implicit hc: HeaderCarrier): Future[GetEstateResponse] = {
```
Comment (adamconder) [inline app/uk/gov/hmrc/estates/services/DesService.scala:46]: Method name is trusts.

File: app/uk/gov/hmrc/estates/services/DesService.scala [hunk 4]
```diff
@@ -38,8 +43,33 @@ class DesService @Inject()(val desConnector: DesConnector) {
...
+    Logger.debug("Retrieving Estate Info from DES")
+    Logger.info(s"[DesService][refreshCacheAndGetTrustInfo] refreshing cache")
+
+    repository.resetCache(utr, internalId).flatMap { _ =>
+      desConnector.getEstateInfo(utr).map {
+        case response: GetEstateProcessedResponse =>
+          repository.set(utr, internalId, Json.toJson(response)(GetEstateProcessedResponse.mongoWrites))
```
Comment (adamconder) [inline app/uk/gov/hmrc/estates/services/DesService.scala:53]: Why don't we wait for the future here? 🤔

File: app/uk/gov/hmrc/estates/transformers/amend/AmendPersonalRepTransform.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.estates.transformers.amend
+
+import play.api.libs.json._
+
+trait AmendPersonalRepTransform {
+  def setPersonalRep(input: JsValue, newPersonalRepDetails: JsValue): JsResult[JsValue] = {
+    val personalRepPath = (__ \ 'details \ 'estate \ 'entities \ 'personalRepresentative)
```
Comment (adamconder) [inline app/uk/gov/hmrc/estates/transformers/amend/AmendPersonalRepTransform.scala:23]: Can you double check this? Is this the array which contains perRepInd perRepOrg objects?

File: it/uk/gov/hmrc/repositories/CacheRepositorySpec.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          val retrieved = repository.get("UTRUTRUTR", "InternalId")
+            .map(_.getOrElse(fail("The record was not found in the database")))
+
+          retrieved.futureValue mustBe data
+
+          dropTheDatabase(connection)
+        }.get
+      }
+    }
+  }
```
Comment (adamconder) [inline it/uk/gov/hmrc/repositories/CacheRepositorySpec.scala:52]: There was a safer implementation that I came up with in the maintain-other-individuals-frontend microservice. Can you bring this over as it means we don't have to do `.get`.

File: app/uk/gov/hmrc/estates/services/DesService.scala (hunk 4 above)
Comment (BriWak) [reply app/uk/gov/hmrc/estates/services/DesService.scala]: Fixed now

File: app/uk/gov/hmrc/estates/services/DesService.scala (hunk 3 above)
Comment (BriWak) [reply app/uk/gov/hmrc/estates/services/DesService.scala]: Fixed now

File: app/uk/gov/hmrc/estates/transformers/amend/AmendPersonalRepTransform.scala (hunk 5 above)
Comment (BriWak) [reply app/uk/gov/hmrc/estates/transformers/amend/AmendPersonalRepTransform.scala:23]: Yes, I've checked and this is the same as Lead Trustee in trusts so I've used the same implementation

File: it/uk/gov/hmrc/repositories/CacheRepositorySpec.scala (hunk 6 above)
Comment (BriWak) [reply it/uk/gov/hmrc/repositories/CacheRepositorySpec.scala:52]: maintain-other-individuals-frontend has no tests for the repository that I can see?

File: app/uk/gov/hmrc/estates/transformers/register/VariationDeclarationTransform.scala [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  private def putNewValue(path: JsPath, value: JsValue ): Reads[JsObject] =
+    __.json.update(path.json.put(value))
+
+  private def declarationAddress(agentDetails: Option[AgentDetails], responseJson: JsValue) =
+    if (agentDetails.isDefined)
+      agentDetails.get.agentAddress
+    else
+      responseJson.transform((pathToPersonalRep \ 'identification \ 'address).json.pick) match {
+        case JsSuccess(value, _) => value.as[AddressType]
+        case JsError(_) => ???
```
Comment (adamconder) [inline app/uk/gov/hmrc/estates/transformers/register/VariationDeclarationTransform.scala:124]: Unimplemented?

**Pull request merged**